### PR TITLE
✨ Trigger machine pool instance refresh (node rollout) if bootstrap config reference changes

### PR DESCRIPTION
**Pull request by AndiDog**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

We discussed this and alternative solutions at length in https://github.com/kubernetes-sigs/cluster-api/issues/8858 and CAPI/CAPA office hours. Eventually, we turned to a solution that is not surprising for users: if the bootstrap config reference (e.g. `KubeadmConfig`) changes within a `MachinePool`/`AWSMachinePool` combo, i.e. its name changes, new nodes should be rolled out. With this, it is not necessary to parse the bootstrap data (= user data) and detect changes within those scripts (cloud-init, ignition, maybe other formats in the future). Instead, we only need to store the previously-used bootstrap data secret key (namespace+name) in the launch template, and compare on every reconciliation whether that reference changed. For example, if an operating user wants to roll out a change to `KubeadmConfig.spec.files`, i.e. add new files on the nodes, it's now sufficient to create a newly-named `KubeadmConfig`. For my specific usage, this means naming the object `KubeadmConfig/my-config-<hash of its spec>` and the Helm chart upgrade will take care to deploy this new object and delete the old one (which I tested and works perfectly fine).

Testing of this change showed that CAPI+CAPA correctly, and in the correct order without possible race conditions (🤞), render the bootstrap data secret, mark the bootstrap config ready, create a new launch template version and trigger instance refresh (which creates new EC2 instances). The only minor glitch I found is https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4618, but that can be solved separately.

Since reconciliation of launch templates wasn't covered in tests at all, I changed the mock interfaces a little so it could be tested, and added some test cases.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Towards https://github.com/kubernetes-sigs/cluster-api/issues/8858 and https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4071

For the feature to work, users require a CAPI version that includes https://github.com/kubernetes-sigs/cluster-api/pull/8667. Without that, CAPA won't see an updated `MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` to reconcile.

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Trigger machine pool instance refresh (node rollout) if bootstrap config reference changes
```

### CODE REVIEW

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (fiunchinho) [inline pkg/cloud/services/ec2/launchtemplate.go:877]: Do we use blocks like `{ }` on other parts of the code? I'm wondering if we want to use it, or if it improves things. If we don't use it anywhere else, I wouldn't use it.

**Review by cnmcavoy — COMMENTED** [review]
Take a look at the linter errors failing the tests. Overall this looks solid.

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (AndiDog) [reply pkg/cloud/services/ec2/launchtemplate.go:877]: If there are zero tags per category (here: `ec2.ResourceTypeInstance` and `ec2.ResourceTypeVolume`), we just shouldn't add the tag specification. That's why the guard is there. And I think having a block, whether it's just `{ ... }` or `if ... { ... }` around each category, together with a comment above each block, separates the logic somewhat cleanly.

File: pkg/cloud/services/interfaces.go
Comment (AndiDog) [reply pkg/cloud/services/interfaces.go:88]: `EC2Interface` mocks the AWS EC2 API, so `ReconcileLaunchTemplate` (which I moved out) and `ReconcileTags` (out of scope for my PR) should not be in there. If I move the `ReconcileTags` function as well into the new interface, it should become clearer for someone working on the code, particularly if we state that it's for the functionality of the controller as opposed to raw AWS requests. Given that change, what about `MachinePoolReconcileInterface` since those functions are exclusively used in `awsmachinepool_controller.go` and `awsmanagedmachinepool_controller.go`?

Comment (cnmcavoy) [reply pkg/cloud/services/interfaces.go:88]: I think that would clear up my confusion, sgtm.

Comment (AndiDog) [reply pkg/cloud/services/interfaces.go]: Done. Also fixed lint errors.

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (cnmcavoy) [inline pkg/cloud/services/ec2/launchtemplate.go:725]: Can we return here?

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (cnmcavoy) [inline pkg/cloud/services/ec2/launchtemplate.go:731]: is it okay for `launchTemplateUserDataSecretKey` to be potentially nil here, if the above loop fails? Or should we return an error here?

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (AndiDog) [reply pkg/cloud/services/ec2/launchtemplate.go:725]: Done

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (AndiDog) [reply pkg/cloud/services/ec2/launchtemplate.go]: We must be backward-compatible and support cases where the tag is missing by mistake/bug/intervention. So `nil` is fine. I changed this line to explicit `nil` here, and added the `return` for the found case into the loop as you commented above.

File: pkg/cloud/services/ec2/launchtemplate.go
Comment (AndiDog) [reply pkg/cloud/services/ec2/launchtemplate.go:897]: Does the above answer solve your question? Here, `tags` could be empty and we want an `if`, while for instance tags, we now have at least one (the newly added tag used for the feature) and need no `if`.

**Review by cnmcavoy — APPROVED** [review]
/lgtm

**Comment by AndiDog** [comment]
Rebased and only one (obvious to me) test line was conflicted. @cnmcavoy @fiunchinho could you take another look? @richardcase we'd need approval here as well. I'm not sure if this should go into v2.3.x?

**Comment by richardcase** [comment]
> I'm not sure if this should go into v2.3.x?

@AndiDog - as its classed as a feature i'd say it needs to go into v2.4.0 instead. We generally only consider bugs for a patch release.

**Comment by AndiDog** [comment]
@cnmcavoy @richardcase The PR needs LGTM (again after rebase) and approval.

I agree it's a feature, so no need to backport. We have integrated it already in our CAPA fork and it's working.

**Review by cnmcavoy — APPROVED** [review]
/lgtm

**Comment by richardcase** [comment]
This looks good to me. Lets check the e2e passes before merging:

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

**Comment by damdo** [comment]
@AndiDog see for your e2e failure, [see this](https://kubernetes.slack.com/archives/CD6U2V71N/p1705574166982269?thread_ts=1705565476.663379&cid=CD6U2V71N)

**Comment by AndiDog** [comment]
Rebased because that fixed E2E tests in another PR.

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

**Review by fiunchinho — APPROVED** [review]
/lgtm

**Comment by AndiDog** [comment]
Still one E2E failure related to LBs. The test function `createLBService` look strange to me. It expects some status field to be filled immediately or otherwise silently continues.

**Comment by Ankitasw** [comment]
/approve

cc @fiunchinho @cnmcavoy @AndiDog for final review

**Review by richardcase — COMMENTED** [review]
Looks good to me as well

/approve

For the final lgtm:

/assign @cnmcavoy @fiunchinho

**Pull request merged**
